File: exercises/ch02/05/java/Main.java
import static java.lang.System.in;
import static java.lang.System.out;


class Main {

    static int checkDigit(int checksum) {
	return (10 - checksum % 10) % 10;
    }

    static int multiplyIsbnDigit(int digit, int position) {
	if (position % 2 == 0) return 3 * (digit - '0');
	else return digit - '0';
    }

    public static void main(String[] args) throws java.io.IOException {
	int position = 1; // valid positions [1, 12]
	int checksum = 0;
	int digit; // will store values in range [0,255]

	while ((digit = in.read()) != '\n') {
	    if (digit != ' ' && digit != '-') {
		if (position < 13) {
		    checksum += multiplyIsbnDigit(digit, position);
		}
		position++;
	    }
	}

	if (position != 13) {
	    out.print("The number is not valid. The code must be 12 digits long.\n");
	} else {
	    out.print("The check digit is: " + checkDigit(checksum) + ".\n");
	}
    }
}

File: exercises/ch02/05/cpp/main.cpp
#include <iostream>

using std::cin;
using std::cout;

int multiply_isbn_digit(char digit, int position) {
    if (position % 2 == 0) return 3 * (digit - '0');
    else return digit - '0';
}

int main() {
    char character;
    int checksum = 0;
    int position = 1;

    cout << "Input a ISBN code: ";

    character = cin.get();
    while (character != 10) {
        if (character != 32 && character != 45) {
            if (position <= 13) {
                checksum += multiply_isbn_digit(character, position);
            }
            position++;
        }
        character = cin.get();
    }

    if (position != 14) {
        cout << "Error: The length is invalid.\n";
        return 1;
    }

    if (checksum % 10 == 0) cout << " The code is valid.\n";
    else cout << " The code is invalid.\n";

    return 0;
}
